### DISCUSSION TITLE
Sample a mesh - but get an aggregate result - min max etc.

### DISCUSSION THREAD

**Discussion by RichardScottOZ** [pyvista/pyvista#5498] [Q&A] [answered]
e.g. sample a point cloud array onto a Uniform Grid - and if a particular grid cell/voxel contained 3 point cloud points such as 5, 12, 73 if you wanted to get the max it would sample you 73 or the min you would get 5.

Has anyone looked at anything like this in this space?

**Comment by MatthewFlamm** [comment]
I'd probably start by using `find_containing_cells` and then writing my own logic for how to aggregate the data.

**Reply by RichardScottOZ** [reply]
Is there a function for that already?

**Reply by RichardScottOZ** [reply]
ok, didn't realise find_containing_cell could have an array thrown at it, so thanks Matthew

**Comment by RichardScottOZ** [comment, marked as answer]
So something like this, given a point cloud?

```python
origin = (point_cloud.bounds[0],point_cloud.bounds[2],point_cloud.bounds[4])
print(origin)

g = pv.UniformGrid()
g.dimensions = (83,76,100)

g.origin = origin
g.spacing = (50000,50000,100)

indices = resultsmall.find_containing_cell(point_cloud.points)

df = pd.DataFrame({'cell_index': indices, 'point': point_cloud['depth']})
grouped = df.groupby('cell_index').min()

a = np.empty(resultsmall.n_cells)
a[:] = np.nan

newindex = grouped.reset_index()['cell_index'][1:].values.tolist()
a[newindex] = grouped.reset_index()['point'][1:].values

resultsmall['adepth'] = a
```

![image](https://github.com/pyvista/pyvista/assets/72196131/ba88221d-1365-4372-94cd-64687ddedecb)

![image](https://github.com/pyvista/pyvista/assets/72196131/f61d743a-8d69-446f-a451-0a5d8f6fbaba)

![image](https://github.com/pyvista/pyvista/assets/72196131/14f1c90b-a1fe-469d-84cd-38fb073fe2d9)

For some reason, not plotting though

```python
resultsmall.plot(scalars='adepth',show_grid=True)
```

![screenshot](https://github.com/pyvista/pyvista/assets/72196131/24b1cac2-4a22-4c61-9e61-027d07702bca)

**Reply by RichardScottOZ** [reply]
as this seems fine (I clipped it to -1000 when I made it so that spike makes sense) 
![image](https://github.com/pyvista/pyvista/assets/72196131/c2a1a108-6ab1-45d1-af5a-2fd05bf04403)

**Reply by MatthewFlamm** [reply]
You can't `sample` from a point cloud in most cases, `sample` looks inside the cells.  However, you can use the new `snap_to_closest_point` option to grab the value from the nearest point.  Doesnt sound like what you want though.

**Reply by MatthewFlamm** [reply]
The first code doesnt match the screen shots, so I'm also a bit lost.

**Reply by RichardScottOZ** [reply]
No, want exact locations basically...nearest point could be 800km away, that sort of thing...that would be useful in other cases

Yes, in the past had voxelised to sample ftom

original point cloud, make a grid, sample = resultsmalll which doesnt work in the above..is a bit out of order

**Reply by MatthewFlamm** [reply]
I'm still a bit lost, if you have a uniform grid that is down sampled, you are by definition not going to get exact locations regardless.  If you want smoothed data using the point cloud data, you can try `interpolate` or the `find_containing_cell` idea.  I suspect that `sample` is not what you want.

**Reply by RichardScottOZ** [reply]
No, don't want an interpolation. e.g. a scientific sample done at a given spot, some voxels might have multiple samples -  large numbers will have none - interpolation to the other areas is pretty meaningless.

I just used a downsampled version for speed of testing the workflow - will find out what some of compute is needed to test it at a much higher resolution.

'Exact' as in 'within that box' as opposed to 'interpolated from possibly 800km away.

**Reply by MatthewFlamm** [reply]
Confusingly, `interpolate` in PyVista does not interpolate values like this.  It smooths data over nearby points in a point cloud.  `sample` is more like the interpolate you are referring to.

**Reply by RichardScottOZ** [reply]
As I think I understand it, sample interpolates on what is actually in the cell?

**Reply by MatthewFlamm** [reply]
Yes, and since your point cloud has vertex cells (0D cells) `sample` filter cannot do this. The best it can do is snap to nearest cell.    `interpolate` filter can smooth the data from a point cloud over some distance function.  The `find_containing_cell` approach is another.

**Comment by RichardScottOZ** [comment]
Ok, something with the ghostmasks?

```python
resultsmall.clear_data()
resultsmall

resultsmall['adepth'] = a
resultsmall.plot(scalars='adepth')
```

![image](https://github.com/pyvista/pyvista/assets/72196131/a43b6a4d-b49b-43f3-aee7-cb960321e3de)

![screenshot](https://github.com/pyvista/pyvista/assets/72196131/3659ef81-3ce6-46fd-ab99-b4bd6dce936f)

at least that should be somewhere in the right direction?
